src/irr/asset/CGLIWriter.cpp

```diff
@@ -0,0 +1,327 @@
+		
+			const auto texelBlockByteSize = asset::getTexelOrBlockBytesize(imageInfo.format);
+			const auto channelsAmount = getFormatChannelCount(imageInfo.format);
+			const auto singleChannelByteSize = texelBlockByteSize / channelsAmount;
+			const auto data = reinterpret_cast<const uint8_t*>(image->getBuffer()->getPointer());
+
+			auto getTarget = [&]()
+			{
+				switch (imageViewInfo.viewType)
+				{
+					case ICPUImageView::ET_1D: return gli::TARGET_1D;
+					case ICPUImageView::ET_1D_ARRAY: return gli::TARGET_1D_ARRAY;
+					case ICPUImageView::ET_2D: return gli::TARGET_2D;
+					case ICPUImageView::ET_2D_ARRAY: return gli::TARGET_2D_ARRAY;
+					case ICPUImageView::ET_3D: return gli::TARGET_3D;
+					case ICPUImageView::ET_CUBE_MAP: return gli::TARGET_CUBE;
+					case ICPUImageView::ET_CUBE_MAP_ARRAY: return gli::TARGET_CUBE_ARRAY;
+					default: assert(0);
+				}
+			};
+
+			auto getFacesAndLayersAmount = [&]()
+			{
+				size_t layers, faces;
+				const auto arrayLayers = imageInfo.arrayLayers;
+
+				if (layersFlag)
+					layers = ((arrayLayers - 1) / 6) + 1;
+				else
+					layers = 1;
+				if (facesFlag)
+					faces = ((arrayLayers - 1) % 6) + 1;
+				else
+					faces = 1;
+
+				return std::make_pair(layers, faces);
+			};
+
+			auto gliFormatAndSwizzles = getTranslatedIRRFormat(imageViewInfo);
+			gli::target gliTarget = getTarget();
+			gli::extent3d gliExtent3d = {imageInfo.extent.width, imageInfo.extent.height, imageInfo.extent.depth};
+			size_t gliLevels = imageInfo.mipLevels;
+			std::pair<size_t, size_t> layersAndFacesAmount = getFacesAndLayersAmount();
+
+			gli::texture texture(gliTarget, gliFormatAndSwizzles.first, gliExtent3d, layersAndFacesAmount.first, layersAndFacesAmount.second, gliLevels, gli::texture::swizzles_type{ gliFormatAndSwizzles.second[0], gliFormatAndSwizzles.second[1], gliFormatAndSwizzles.second[2], gliFormatAndSwizzles.second[3] });
+
+			auto getFullSizeOfLayer = [&](const uint16_t mipLevel) -> uint64_t
+			{
+				auto region = image->getRegions().begin() + mipLevel;
+				const auto width = region->bufferRowLength == 0 ? region->imageExtent.width : region->bufferRowLength;
+				const auto height = region->bufferImageHeight == 0 ? region->imageExtent.height : region->bufferImageHeight;
+				const auto depth = region->imageExtent.depth;
+				return width * texelBlockByteSize * height * depth;
+			};
+
+			for (auto region = image->getRegions().begin(); region != image->getRegions().end(); ++region)
+			{
+				const auto ptrBeginningOfRegion = data + region->bufferOffset;
+				const auto layerSize = getFullSizeOfLayer(region->imageSubresource.mipLevel);
+
+				for (uint16_t layer = 0; layer < imageInfo.arrayLayers; ++layer)
+				{
+					const uint16_t gliLayer = layersFlag ? layer / 6 : 0;
+					const uint16_t gliFace = facesFlag ? layer % 6 : 0;
+					memcpy(texture.data(gliLayer, gliFace, region->imageSubresource.mipLevel), ptrBeginningOfRegion + (layer * layerSize), layerSize);
+				}	
```

Comment:
you're doing it all wrong, a region does not contain all the mip levels also layers are counted differently for non-cubemap textures.

Also you can't do it with a memcpy (well you can but at least one `memcpy` per image-row) because you need to honor the region's defined `bufferRowLength` and `bufferImageHeight` to increment the source pointer properly.

Reply:
My single region contains all layers and faces on certain single mip level, but I forgot about 2d 1d arrays, so I have to fix handling layers amount for sure

Reply:
but not sure what about `bufferImageHeight `

Reply:
This will not work
```cpp
memcpy(texture.data(gliLayer, gliFace, region->imageSubresource.mipLevel), ptrBeginningOfRegion + (layer * layerSize), layerSize);
```

if your `bufferImageHeight!=height` or `bufferImageWidth!=width`, you can't copy more than a row at once.

Reply:
Fixed, but I will touch it yet a little tomorrow